rotkehlchen/chain/evm/decoding/zerox/decoder.py

```diff
@@ -146,43 +185,57 @@ def _decode_swap(
                 hex_or_bytes_to_address(_log.topics[2]) == self.flash_wallet_address
             ):
-                self._update_send_and_receive_events(send_event=send_event)
-
-        summed_send_event, summed_receive_event = (
-            self.base.make_event_next_index(
-                tx_hash=transaction.tx_hash,
-                timestamp=transaction.timestamp,
-                asset=asset,
-                balance=Balance(),  # Empty init. Populated below
-                event_type=HistoryEventType.TRADE,
-                event_subtype=subtype,
-                location_label=transaction.from_address,
-            ) for asset, subtype in (
-                (sent_asset, HistoryEventSubType.SPEND),
-                (received_asset, HistoryEventSubType.RECEIVE),
-            )
+                self._update_send_receive_fee_events(send_event=send_event)
+
+        self._merge_similar_events(
+            decoded_events=decoded_events,
+            send_events=list(send_address_to_events.values()),
+            receive_events=list(receive_address_to_events.values()),
         )
 
-        for send_event in send_address_to_events.values():
-            summed_send_event.balance += send_event.balance  # add sent balances
-            decoded_events.remove(send_event)
-        for receive_event in receive_address_to_events.values():
-            summed_receive_event.balance += receive_event.balance  # add received balances
-            decoded_events.remove(receive_event)
+        return decoded_events
 
-        self._update_send_and_receive_events(
-            send_event=summed_send_event,
-            receive_event=summed_receive_event,
-        )
+    def _decode_meta_tx_swap(self, context: 'DecoderContext') -> DecodingOutput:
+        """Decodes the swap event from the 0x router contract via executeMetaTransactionV2."""
+        if context.tx_log.topics[0] != META_TX_ZEROX or context.tx_log.address != self.router_address:  # noqa: E501
+            return DEFAULT_DECODING_OUTPUT
 
-        decoded_events.append(summed_send_event)
-        decoded_events.append(summed_receive_event)
-        maybe_reshuffle_events(
-            ordered_events=[summed_send_event, summed_receive_event],
-            events_list=decoded_events,
+        send_events, receive_events, fee_event = [], [], None
+        for event in context.decoded_events:
+            if ((
+                event.event_type == HistoryEventType.TRADE and
+                event.event_subtype == HistoryEventSubType.SPEND
+            ) or (
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.address == self.flash_wallet_address
+            )) and event.location_label is not None and self.base.is_tracked(event.location_label):  # type: ignore  # it's a checksum address here
+                send_events.append(event)
```

Comment:
for this last case can't you move it to the fee event and check if ` event.address == self.flash_wallet_address`. If so it is a send_event otherwise it is a fee event? This way you avoid extra checks